examples/sync/src/protocol.rs

```diff
@@ -0,0 +1,679 @@
+//! Network protocol definitions for ADB sync.
+//!
+//! This module defines the network protocol used for ADB synchronization between
+//! clients and servers. It includes message types, error handling, and validation
+//! logic for safe network communication.
+//!
+//! The protocol supports:
+//! - Getting server metadata (database size, target hash, operation bounds)
+//! - Fetching operations with cryptographic proofs
+//! - Comprehensive error handling
+
+use bytes::{Buf, BufMut};
+use commonware_codec::{EncodeSize, Error as CodecError, Read, ReadExt, ReadRangeExt as _, Write};
+use commonware_cryptography::sha256::Digest;
+use commonware_runtime::{Sink, Stream};
+use std::num::NonZeroU64;
+use thiserror::Error;
+
+/// Protocol version identifier.
+pub const PROTOCOL_VERSION: u8 = 0;
+
+/// Maximum message size in bytes (10MB).
+pub const MAX_MESSAGE_SIZE: usize = 10 * 1024 * 1024;
+
+/// Read a length-prefixed message from a stream.
```

Comment:
You can use `framing` for this: https://github.com/commonwarexyz/monorepo/blob/58aa05b738986567b43507243ef2451feca27eab/stream/src/utils/codec.rs#L5-L50